velox/dwio/parquet/tests/reader/ParquetTableScanTest.cpp

```diff
@@ -820,13 +832,15 @@ TEST_F(ParquetTableScanTest, timestampInt96Dictionary) {
   WriterOptions options;
   options.writeInt96AsTimestamp = true;
   options.enableDictionary = true;
+  options.parquetWriteTimestampUnit = TimestampUnit::kMicro;
```

Comment:
Int96 consists of days and nanos, so I wonder if it makes sense to write with the default nano precision. Thanks.

Reply:
Actually, per https://github.com/apache/arrow/issues/36005, we could NOT write a Timestamp in nanoseconds. In fact, if you look at the Timestamp data changes in `testInt96TimestampRead` in this PR, I only specify at most in the microseconds, because if we have `1999-12-08 13:39:26.123456789` while both `parquetWriteTimestampUnit` and the reading precision are in nanoseconds, we could only get `1999-12-08 13:39:26.123456000`, where the nanosecond part gets truncated.

Also, I'm not sure how the dictionary encoding works in Parquet. Even w/ `enableDictionary` is `true` in `timestampInt96Dictionary`, it does not call `WriteArrowDictionary`. Is it related to the test data in `testInt96TimestampRead` where the dictionary encoding does not reduce data size so it falls back to plain encoding?
https://github.com/facebookincubator/velox/blob/b43795025c6f12852a6eb91db076f6b9cd7f0001/velox/dwio/parquet/writer/arrow/ColumnWriter.cpp#L1631-L1632

Reply:
Thank you for sharing the Arrow's limitation!

> Is it related to the test data in testInt96TimestampRead where the dictionary encoding does not reduce data size so it falls back to plain encoding?

I assume it does not fall back, and I printed some log at the below point and verified it entered the dictionary reading.

https://github.com/facebookincubator/velox/blob/7a023214aa4033eba625cc9d965c7db65e35f803/velox/dwio/parquet/reader/PageReader.cpp#L380

About how the dictionary encoding works, I notice 'DictEncoderImpl' will be used if enableDictionary is true.

https://github.com/facebookincubator/velox/blob/7a023214aa4033eba625cc9d965c7db65e35f803/velox/dwio/parquet/writer/arrow/Encoding.cpp#L4133-L4139

Reply:
Thank you a lot for pointing to PageReader. It saved me a day to figure out why Int64 Timestamp in the dictionary encoding case gets read as a vector of `int64`, instead of Velox `Timestamp` in 128-bit!

Fyi, I just found the place where Arrow timestamp in nanoseconds gets truncated into microseconds.
https://github.com/facebookincubator/velox/blob/ac5c15eb7627b42f8d41023215d1d4a7e69d45aa/velox/dwio/parquet/writer/arrow/ColumnWriter.cpp#L2584-L2595

Reply:
Thank you for sharing the information! Is truncating limited to versions of Parquet 1.0 and 2.4? It appears that Velox uses the Parquet-2.6 version. https://github.com/facebookincubator/velox/blob/b44ffc9e9e85e3a0d21ec72f2e463ff17be011ee/velox/dwio/parquet/writer/arrow/Properties.h#L326


I also notice the below exception in the Spark query runner if without https://github.com/facebookincubator/velox/pull/11698/files#diff-04bd90dfc1fe1e50c098001f60ae4bdca4b61e662918661a141c7db948b28d47R54. It seems nano int64 timestamps can be generated.

```
Caused by: org.apache.spark.sql.AnalysisException: Illegal Parquet type: INT64 (TIMESTAMP(NANOS,false)).
```